core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java

```diff
@@ -0,0 +1,351 @@
+    verifyNoMoreInteractions(mockSubchannel);
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterStateChange() throws Exception {
+    when(mockHelper.createSubchannel(any(EquivalentAddressGroup.class), any(Attributes.class)))
+        .then(new Answer<Subchannel>() {
+          @Override
+          public Subchannel answer(InvocationOnMock invocation) throws Throwable {
+            return createMockSubchannel();
+          }
+        });
+    loadBalancer.handleResolvedAddresses(Lists.newArrayList(servers.keySet()), Attributes.EMPTY);
+    verify(mockHelper, times(1)).updatePicker(isA(Picker.class));
+
+    Subchannel subchannel = loadBalancer.getSubchannels().iterator().next();
+    AtomicReference<ConnectivityStateInfo> subchannelStateInfo = subchannel.getAttributes().get(
+        STATE_INFO);
+
+    assertThat(subchannelStateInfo.get()).isEqualTo(ConnectivityStateInfo.forNonError(IDLE));
+
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+
+    assertThat(subchannelStateInfo.get()).isEqualTo(
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forTransientFailure(Status.UNKNOWN));
+
+    assertThat(subchannelStateInfo.get()).isEqualTo(
+        ConnectivityStateInfo.forTransientFailure(Status.UNKNOWN));
+  }
+
+  @Test
+  public void pickerRoundRobin() throws Exception {
+    Subchannel subchannel = mock(Subchannel.class);
+    Subchannel subchannel1 = mock(Subchannel.class);
+    Subchannel subchannel2 = mock(Subchannel.class);
+
+    Picker picker = new Picker(Collections.unmodifiableList(
+        Lists.<Subchannel>newArrayList(subchannel, subchannel1, subchannel2)), null);
+
+    assertEquals(subchannel,
+        picker.pickSubchannel(Attributes.EMPTY, new Metadata()).getSubchannel());
+    assertEquals(subchannel1,
+        picker.pickSubchannel(Attributes.EMPTY, new Metadata()).getSubchannel());
+    assertEquals(subchannel2,
+        picker.pickSubchannel(Attributes.EMPTY, new Metadata()).getSubchannel());
+    assertEquals(subchannel,
+        picker.pickSubchannel(Attributes.EMPTY, new Metadata()).getSubchannel());
+  }
+
+  @Test
+  public void pickerEmptyList() throws Exception {
+    Picker picker = new Picker(Lists.<Subchannel>newArrayList(), Status.UNKNOWN);
+
+    assertEquals(null, picker.pickSubchannel(Attributes.EMPTY, new Metadata()).getSubchannel());
+    assertEquals(Status.UNKNOWN,
+        picker.pickSubchannel(Attributes.EMPTY, new Metadata()).getStatus());
+  }
+
+  @Test
+  public void nameResolutionErrorWithNoChannels() throws Exception {
+    loadBalancer.handleNameResolutionError(Status.NOT_FOUND.withDescription("nameResolutionError"));
+    verify(mockHelper).updatePicker(pickerCaptor.capture());
+    LoadBalancer2.PickResult pickResult = pickerCaptor.getValue().pickSubchannel(Attributes.EMPTY,
```

Comment:
> (subchannels and status, you may want to save the list in picker just for testing)

Thought about that, but I really don't like adding code in impl just to make it more stable. I "worked" around this just by iterating over the list X times and collecting results.